### DISCUSSION TITLE
Rows and Columns and Sorting and Hashing (oh my!)

### DISCUSSION THREAD

**Discussion by hawkfish** [duckdb/duckdb#1777] [Ideas]
@lnkuiper Maybe this is a good place to discuss sorting and other row-oriented issues?

The layout I have settled on for hash tables is column validity mask followed by data fields. This makes the scan APIs simpler, but I do worry about wasted space for the masks caused by alignment issues.

The big change I made to the existing code was to stop updating pointer columns when scanning across a row and instead use a column of pointers to the start of each row, a column number and an offset that gets updated during a scan.  The offset starts at the end of the mask and can be updated using the column type. This avoids unnecessary cache writes during read operations (in fact, it makes scanning a read-only operation) and all columns can access the validity mask via their column index.

I'm thinking now that these scans could even be templatised on the physical types, but the types of operations (sort, scatter, gather) might be too varied to make this useful.

If you want to see the current version of this, have a look at [this PR](https://github.com/duckdb/duckdb/pull/1775).

**Comment by lnkuiper** [comment]
> The layout I have settled on for hash tables is column validity mask followed by data fields.

This is great, I have done the same for the row-based sorting.

> I do worry about wasted space for the masks caused by alignment issues.

Correct me if I'm wrong, but the wasted space is in the order of bits per row, right? (max 7 bits wasted per row) I wouldn't worry about it!

> The offset starts at the end of the mask and can be updated using the column type.

Again, same for sorting, which I conveniently copied from the JoinHashTable

> I'm thinking now that these scans could even be templatised on the physical types, but the types of operations (sort, scatter, gather) might be too varied to make this useful.

I think this could be useful. I think we can settle on a specific format for the serialization of rows, and have different operations on this format.

Have you taken a look at [RowDataCollection](https://github.com/duckdb/duckdb/blob/master/src/common/types/row_data_collection.cpp)? I have [serialization](https://github.com/duckdb/duckdb/blob/master/src/common/types/row_data_collection.cpp#L303) and [deserialization](https://github.com/duckdb/duckdb/blob/master/src/common/types/row_data_collection.cpp#L696). What's also nice about this is that it already supports row-wise serialization of lists/structs.

I am sure that JoinHashTable could use RowDataCollection, which would unify their formats, making use of the same row-wise validity mask. I haven't looked into the aggregate hash table, but from your PR it seems that this is quite similar, but I cannot say for sure since I don't know much about it. Can you take a look at RowDataCollection? @hawkfish

**Reply by hawkfish** [reply]
I will have a good look later, but this looks like the other half of the puzzle. I especially like that `STRUCT`s and `LIST`s are handled and of course the sorting is important.

My own noodling on this is leading me to a structure I am calling `RowChunk` (or maybe `RowView` or `RowLayout`?) It manages all the fiddly details about row layout including validity masks, column types and locations and offsets. Clients then only need to know about column numbers. It also manages aggregate states for aggregation and potentially `GroupJoin` operators

The idea is that it would have  a "schema" consisting of a set of data types and a list aggregate functions. Once initialised, it has all the offsets, padding and masking dealt with for the client. The client then uses it by just using `Reference` to set the pointer columns.  Operations might include:

* `Initialize` with a list of `LogicalType`s and `AggregateFunction`s. This will compute all the layouts and sizing for the rows.
* `Scatter` (not the best name). Takes a `DataChunk` and writes new data into the columns. It also sets the validity masks and calls the `initialise` method on any aggregate states.
* `Gather` (again, not the best name). Extracts data for a single `Vector` or an entire `DataChunk`
* `Update`, `Combine`, `Finalize` Apply aggregate operations to single aggregates or all of them. `Update` Reads from a `DataChunk` and `Finalize`writes to a `Vector`
* Row level operations (templated). An `Operation` method gets N typed values and computes the result into an output `Vector` or into another field (e.g., `AdvancePointers`). Unary, binary, etc.
* `Partition` Uses a (templated) predicate to build two `SelectionVector`s.
* `Swap`(?) Swaps the physical contents of two rows.

Sorting was something I was not really clear on, nor were compound data types. Hopefully we can mash these two ideas together - I don't think they are all that dissimilar.

You mention that alignment is not a big problem (7 bits) but my current implementations are a lot more wasteful (63!). One of the things I am hoping to get out of this data structure is being able to check on things like that in one place instead of scattered all over the code. I also need to see how you are doing row comparisons and how the validity mask location affects sorting.

[Here's what I typed in](https://github.com/duckdb/duckdb/files/6516312/row_chunk.txt). It is still very raw and I only thought of the aggregate states when I woke up in the middle of the night(!)

**Reply by hawkfish** [reply]
After having a good look at `RowDataCollection` I think I see the flaw in our plan. The sort code uses particular representations that can be sorted, which requires reversing endianness on some hardware. The hash table representations are native so that they can be extracted quickly. These requirements seem fundamentally incompatible?

**Reply by Mytherin** [reply]
The amount of bytes stored should still be the same, no? Perhaps this can be resolved by a template type that determines which transformations should be performed on the different data types (for hash tables this would be "none", for sort this would contain the endianness swaps and similar).

**Reply by lnkuiper** [reply]
> Here's what I typed in. It is still very raw and I only thought of the aggregate states when I woke up in the middle of the night(!)

Ha, I love it when these productive rushes occur in the middle of the night!

I think I should clarify `RowDataCollection`, as well as the sorting some more, since it does more than serialising to a sortable representation. When data comes in at `PhysicalOrder::Sink`, the sorting columns are grabbed, and serialised to a fixed-size radix-sortable (`memcmp`-able) row format, using the same code as the ART index. This happens in SerializeVectorSortable. If there are variable size sorting columns, only a prefix of these is serialised. This function should only be used in `PhysicalOrder`.

The variable size sorting columns are then also serialised in full (in native format), to be able to break ties if their prefix is the same. Because these rows have different sizes, we cannot simply jump into the block of data using e.g. `entry_idx * entry_size`. In order to do this, another block of data is kept with offsets of each entry. The offsets are computed by first calling `RowDataCollection::ComputeEntrySizes`, and converting the entry sizes to offsets later: now we can jump into the block of data to get a specific row.

The payload is then also serialized, in native format. Again, if there are no variable sized columns, we simply serialise, and we can jump into the data using `entry_idx * entry_size`, if not, we use the same trick as for the variable size sorting columns, and keep an offset block. The payload has a validity bitmask in front of each row, of size `(num_cols + 7) / 8)`, thus wasting only 7 bits, depending on the number of columns. Here's a small sketch:
```
payload
row 0: | validity bits | column 0 | column 1 | ... | column N|
row 1: | validity bits | column 0 | column 1 | ... | column N|
row 2: | validity bits | column 0 | column 1 | ... | column N|
row 3: | validity bits | column 0 | column 1 | ... | column N|

payload offsets:
| 0 | offset of row 1 | offset of row 2 | offset of row 3 | end of row 3 |
```
As you can see, the payload offsets have 1 more entry than the payload itself, to store the end of the last entry. Of course, the 0 at the start could have been omitted, but that would require a lot of `if (entry_idx == 0)` everywhere, which I did not want.

When re-ordering rows after the initial radix sort, or when copying rows during the merge sort, you can easily compute the size of a row using `idx_t size_of_row_N = offsets[N + 1] - offsets[N]`.

I think that this payload representation can be used at least for the `JoinHashTable`, but again, I'm not sure about the aggregates, because I don't know much about them.

I would also like to add that in the case where everything fits in memory, you do not need to have this offset block, and could possibly add a string to the `StringHeap` and just serialise a pointer to this string, giving you fixed-size entries again. This is what `JoinHashTable` does at the moment, but this creates a problem with `LIST` and `STRUCT`, since we do not have e.g. `ListHeap` or `StructHeap`.

I understand that carrying around this block of offsets can be quite cumbersome, and undesirable in some situations, but I don't see any other way of being able to jump into a block of variable sized rows.

I hope this clears `RowDataCollection` up a bit, but perhaps it is easier to hop into a Zoom call. You're probably asleep right now, but I can chat between 15:00 and 16:00 CET if you'd like.

**Reply by hawkfish** [reply]
I was thinking a "storage template" might do the trick. Right now, the hash table layout uses templated `Load` and `Store` for unaligned access to values and the sorting layout has something similar. If they were wrapped in a `struct` (along with `Size` and maybe `Align` methods) , then the storage model could be abstracted away. Endianness could be a parameter for sorting template instantiation.

The other issue is `NULL` storage. The sorting layout puts the `NULL` flag in a byte at the front of the value, but having is using a `ValidityMask` at the front. I expect the storage template could manage that, but it needs to be handled.

**Reply by hawkfish** [reply]
@lnkuiper Sorry I didn't get a notification about your posting and I only just saw it after I posted another reply... I think Monday at that time would work for me, but unfortunately Monday is my busy day next week, so I can't go much later.

**Reply by hawkfish** [reply]
Small detail: How does having the validity bits at the front of the block work with sorting? Wouldn't that make minor sort columns `NULL` semantics override more major sort columns?

**Comment by hawkfish** [comment]
I've been playing around some more with a class I call `RowLayout`. It could eventually be templatised for different formats (e.g, validity masks or header bytes), but first I'm working on replacing all the *ad hoc* pointer arithmetic in the aggregation and join hash tables with this object.

Extending this might look like a `RowOperations` collection analogous to `VectorOperations`. The methods I mentioned above would be organised in various files (aggregation, scatter/gather, etc.). This would hopefully simplify and unify the join and aggregation code and make it easier to support other row-oriented operators such as `GroupJoin`, `IEJoin`, `Order`, `TopN` and `Window`.

For sorting, this would just require a different format template, The existing sorting code could then become another `RowOperations` module, along with maybe heaps?

**Reply by hawkfish** [reply]
Another important part of all this is adding the ability to aggregate and join on `LIST`s and `STRUCT`s. I thought there was an issue filed on this but I can't find it. In any case, the sorting code's ability to serialise these into a row structure seems vital to solving that problem.

**Reply by lnkuiper** [reply]
I think this is a great idea, abstracting away from all the ad-hoc validity bits and rows. But will it be possible to have variable size columns without having to store them in a heap with this approach? This is something that is crucial to out-of-core sorting

**Reply by hawkfish** [reply]
The aggregation/join hash tables only keep strings and they store them in a single string heap. It sounds like the variable sized columns for sorting (including `LIST`s and `STRUCT`s) are kept in a second variable-sized block of data? This is important because the approach I was taking was trying to externalise the storage and only model the layout. With variable-sized data that can only be known at insert time, this will no longer work. Which explains why `RowDataCollection` has a `BufferManager`. This would also be needed for out-of-core hash tables, so it is a common feature.

**Comment by hawkfish** [comment]
@lnkuiper I have set up #1791 to track work on this.

**Comment by hawkfish** [comment]
@lnkuiper Now that I have unified the join and aggregation hash tables, I've started thinking about variable size data structures. Right now, the code just uses a `StringHeap` but I'm wondering if the first step would be to replace it with a `RowDataCollection`.  Even if we don't use all the sorting capabilities, the size computation and column serialisation might be useful.

**Reply by hawkfish** [reply]
It looks like the only reason that the serialise code needs to be members is because of `is_little_endian`. That seems like it could be a class member that is set at system startup? Then we could factor out the sorting representation from the buffer management. Once that is done, then some of the serialisation code for nested types would become polymorphic and we could share it between the hash tables and the sorting code?

**Reply by hawkfish** [reply]
The deserialisation code doesn't seem to use it.

**Reply by lnkuiper** [reply]
> It looks like the only reason that the serialise code needs to be members is because of `is_little_endian`. That seems like it could be a class member that is set at system startup?

Indeed! Otherwise they could be static.

> Then we could factor out the sorting representation from the buffer management.

What do you mean by this? It is important that the sorting representation can stored in buffers, so that we can have out-of-core sorting. If you mean that we can abstract away from the buffer management in the serialization so that it can be used in other places, then I agree. However it should still be possible to store it in buffers.

> The deserialisation code doesn't seem to use it.

Indeed. The radix-sortable data is discarded after the payload is sorted. There is some data duplication going on there: some columns are serialized in both the radix format and the payload.

**Reply by hawkfish** [reply]
I thought the data was read out after sorting?

**Reply by hannes** [reply]
Also, the endianness is known at compile time, no?

**Reply by hawkfish** [reply]
Yes and I was wondering if we could make the computation a `constexpr`?

**Comment by hawkfish** [comment]
I think this may be easier than I thought. For nested types, I think we can just use the existing code in `RowDataOperations` to scatter nested vectors into buffers and then deserialise it for match and gather operations. Nested types will use a two-element struct that contains the pointers to the data locations and the validity masks, and extracting a column just requires extracting these two arrays from the two row values and deserialising. If any of the validity pointers are `nullptr` then there are none.
